pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ImmutableList.java

```diff
@@ -0,0 +1,542 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.ast;
+
+import java.lang.ref.SoftReference;
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+
+/**
+ * Classical immutable linked list representing an ordered collection.
+ * This implementation has constant-time prepend, size, head and tail
+ * operations. Most other operations are linear-time, but caching helps
+ * reduce the overhead of some common operations (reverse). Allows
+ * structural sharing of the tail, which makes many operations constant-
+ * or zero-memory cost.
+ *
+ * <p>This implementation is designed for {@link JavaQualifiedName}, to
+ * represent package and class hierarchies. It is not meant as a general-
+ * purpose implementation.
+ *
+ * @param <E> Element type
+ *
+ * @author Clément Fournier
+ * @since 6.1.0
+ */
+public interface ImmutableList<E> extends Iterable<E> {
```

Comment:
why do we need our own inmutable list implementation? What's wrong with `Collections.unmodifiableList`?

Reply:
`Collections.unmodifiableList` isn't persistent, and doesn't have a free prepend or append operation. I feel like an immutable list like what's found in functional languages is perfect for the problem, and especially for the visitor (next PR). It's good for clarity and conciseness of code, clarity of interface, and performance too. I implemented what I was missing. 

If using this interface instead of List for the return type of a few methods doesn't bother you in itself, then the factories are private so the only dependable code we introduce is the interface, which is in itself very thin. Otherwise, well we could use util.List, but I think it would feel like a workaround...

Reply:
I'm just very cautious about including (and maintaining!) ~600 LoCs for a basic data structure for which Java has several implementations and decorators.... Even if you want a smaller API just allowing prepends, does it earn us anything at all to manually implement the linked list over just using `LinkedList`?

>  This implementation has constant-time prepend, size, head and tail

`LinkedList`:
* [x] constant-time `addFirst()`
* [x] constant-time `size()`
* [x] constant-time `getFirst()`
* [x] constant-time `getLast()`

> Most other operations are linear-time, but caching helps reduce the overhead of some common operations (reverse)

Equally applicable over a `LinkedList`

> Allows structural sharing of the tail, which makes many operations constant- or zero-memory cost.

`LinkedList`:
* [x] `subList()` provides a "view" of the original list, effectively sharing whatever part of the list you want

Even if we keep the `ImmutableList<E>` interface as API, most of the current implementation is just a `LinkedList` when a thin wrapper over  `LinkedList` suffices; so I'd rather just use it than reinvent the wheel

Reply:
The interface of a LinkedList is very different from the interface of that list: `void addFirst(E)` vs `ImmutableList<E> prepend(E)`. A LinkedList can add an element to a list in o(1), while an immutable list can *create a new list* with one more element in o(1) (time and space).  We don't want to add elements fast, we want to create a new, bigger collection fast, and never add elements. This is something you only get with o(n) time and space in java, because you have to copy the old list each time

Besides subList cannot be used for proper structural sharing. You could use it for a `tail` operation but not for a `prepend`. From the subList doc:
> The semantics of the list returned by this method become undefined if the backing list (i.e., this list) is structurally modified in any way other than via the returned list. (Structural modifications are those that change the size of this list, or otherwise perturb it in such a fashion that iterations in progress may yield incorrect results.)

I can't see a way to wrap a LinkedList to make it conform to that interface, without creating a full new LinkedList at each prepend operation, which is exactly what I was trying to avoid. 

I understand the concern about not reimplementing the wheel because it's hard to maintain, but the Java standard lib lacks immutable and persistent collections, which is a shame. Alternatively, I found a light [3rd party lib](https://pcollections.org/) that implements some basic ones, and is very compatible with the standard Collections framework. It's less than 20 classes, and I think it could be useful in other places -- some algorithms are better expressed with them. I'd much rather use that than keep the current full implementation -- i just didn't bother to search

Reply:
I'd rather not add a dependency for a single class.

I'm still hesitant, specially given you are only using an iterator + the ability to skip the first element on this PR, but I'll give you benefit of the doubt to see what's in follow up PRs. I hope you can convince me, I'd hate to see us roll it back in 7.0.0

However, I still strongly believe this class should not be part of this package.